### PR TITLE
Lab 02. File Storage console application

### PR DESCRIPTION
**Pull request by Zhunusau**
# FileStorage .Net Сonsole Application
This is a console application that works on the command line Windows (CMD). The application was made using .Net Framework 4.6.2.

The application starts with 
>  login: **admin**  
> pasword: **admin**
## SonarQube
http://192.168.8.119:9000/dashboard?id=Ruslan-Zhunusau

### CODE REVIEW

**Comment by a-yandulski** [comment]
@Zhunusau, please add description and meaningful name of the PR. Just "Lab 02" is not enough. 
It would be good to see links to the SonarQube and Lab pages in the description.

**Comment by a-yandulski** [comment]
image

Size units - kb, Mb, Gb ???
Extension - dot should be removed as it's just a separator

**Comment by a-yandulski** [comment]
image


image

Expected to see downloaded file as **role.csv**. This is a bug.

**Comment by a-yandulski** [comment]
image

Not user friendly. Display allowed file size within the error message.

**Comment by a-yandulski** [comment]
image

The program should not exit.
File name is incorrect

**Comment by a-yandulski** [comment]
image


Storage size is not recalculated after **upload** \ **remove** operations. This is a bug.

**Comment by a-yandulski** [comment]
image

According to the description of  additional task 2 - there should be:
 - Native support of JSON format
 - Plugin support of XML and YAML

The task is not completed.

**Comment by a-yandulski** [comment]
image

Looks like export \ plugin functionality does not work

File: lab-file-storage/FileStorage/JsonPlugin/JsonSerializer.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            return "json";
+        }
+
+        public void Export(Dictionary<string, MyFile> dictionary, string path)
+        {
+            try
+            {
+                var serializer = new Newtonsoft.Json.JsonSerializer();
+                serializer.Converters.Add(new JavaScriptDateTimeConverter());
```
Comment (a-yandulski) [inline lab-file-storage/FileStorage/JsonPlugin/JsonSerializer.cs]: DateTime should be serialized to a string representation. The following is wrong for this application:
image

Also, would be nice to see an indented json

File: lab-file-storage/FileStorage/XmlPlugin/Contract.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+﻿using System.Collections.Generic;
+using System.Runtime.Serialization;
+using FileStorage.BLL.Models;
+
+namespace XmlPlugin
+{
+    [DataContract]
+    public class Contract
```
Comment (a-yandulski) [inline lab-file-storage/FileStorage/XmlPlugin/Contract.cs]: Strange name. The name of the data contract should correlate with entity name it represents

File: lab-file-storage/FileStorage/XmlPlugin/Contract.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        [DataMember]
+        public Dictionary<string, MyFile> Dictionary { get; set; }
```
Comment (a-yandulski) [inline lab-file-storage/FileStorage/XmlPlugin/Contract.cs]: See comment above. 
Using Dictionary is not very good as it produces  hardly readable xml. Try to use model class to represent the dictionary. Basically output like below should be ok:
image

File: lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            {
+                var assemblyName = AssemblyName.GetAssemblyName(dllFile);
+                var assembly = Assembly.Load(assemblyName);
+                assemblies.Add(assembly);
+            }
+            return assemblies.ToArray();
+        }
+
+        private static Type[] GetIPluginTypes(Assembly[] assemblies)
+        {
```
Comment (a-yandulski) [inline lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs]: Method name is meaningless - the **IPlugin** is already a type

File: lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                {
+                    if (type.IsInterface || type.IsAbstract) continue;
+                    if (type.GetInterface(pluginType.FullName) != null)
+                        pluginTypes.Add(type);
+                }
+            }               
+            return pluginTypes.ToArray();
+        }
+
+        private static IPlugin[] GetIPlugins(Type[] types)
```
Comment (a-yandulski) [inline lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs]: Method name is meaningless. Does this method creates instances  of IPlugin.

File: lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using FileStorage.BLL.Interfaces;
+
+namespace FileStorage.BLL.Services
+{
+    public class PluginManager
```
Comment (a-yandulski) [inline lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs]: It's better to cache plugins in static variable so you don't need to load them on each request. Reflection is an expensive thing

File: lab-file-storage/FileStorage/FileStorage/BLL/Interfaces/IUserService.cs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+﻿namespace FileStorage.BLL.Interfaces
+{
+    public interface IUserService
+    {
+        string Login { get; set; }
+        string UserInfo();
+        bool Validate(string login, string password);
```
Comment (a-yandulski) [inline lab-file-storage/FileStorage/FileStorage/BLL/Interfaces/IUserService.cs]: This method seems to be doing an authentication, so please name it appropriately

File: lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            var pluginType = typeof(IPlugin);
+            var pluginTypes = new List<Type>();
+            foreach (Assembly assembly in assemblies)
+            {
+                if (assembly == null) continue;
+                Type[] types = assembly.GetTypes();
+                foreach (Type type in types)
+                {
+                    if (type.IsInterface || type.IsAbstract) continue;
```
Comment (a-yandulski) [inline lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs]: 'If' statements can be merged into single one

File: lab-file-storage/FileStorage/FileStorage/BLL/Services/StorageService.cs [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+            _files.Remove(file.Key);
+            UsedSize -= file.Value.Size;
+            BinSerialize();
+            return $"The file {fileName} has been removed";
+        }
+
+        public string Export(string path, string name)
+        {
+            IPlugin plugin = null;
+            foreach (var item in _plugins)
```
Comment (a-yandulski) [inline lab-file-storage/FileStorage/FileStorage/BLL/Services/StorageService.cs]: Please use appropriate collection for plugins instead of array

**Comment by a-yandulski** [comment]
Still no description:

image

File: lab-file-storage/FileStorage/FileStorage/Commands/Interfaces/ICommandService.cs [hunk 10]
```diff
@@ -0,0 +1,8 @@
+﻿namespace FileStorage.Commands.Interfaces
+{
+    public interface ICommandService
+    {
+        string[] SplitArguments(string commandLine);
+        void CommandLineExecutor();
```
[the remark this answers is not in the document]
Comment (Zhunusau) [reply lab-file-storage/FileStorage/FileStorage/Commands/Interfaces/ICommandService.cs]: Rename this method

File: lab-file-storage/FileStorage/JsonPlugin/JsonSerializer.cs (hunk 1 above)
Comment (Zhunusau) [reply lab-file-storage/FileStorage/JsonPlugin/JsonSerializer.cs]: fixed this bag

File: lab-file-storage/FileStorage/FileStorage/Configuration/Interfaces/IStorageSection.cs [hunk 11]
```diff
@@ -0,0 +1,10 @@
+﻿namespace FileStorage.Interfaces
+{
+    public interface IStorageSection
+    {
+        string StoragePath { get; set; }
+        string FileData { get; set; }
+        long SizeStorage { get; set; }
```
[the remark this answers is not in the document]
Comment (Zhunusau) [reply lab-file-storage/FileStorage/FileStorage/Configuration/Interfaces/IStorageSection.cs]: renamed names

File: lab-file-storage/FileStorage/FileStorage/Commands/Interfaces/ICommandService.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Zhunusau) [reply lab-file-storage/FileStorage/FileStorage/Commands/Interfaces/ICommandService.cs]: renamed method

File: lab-file-storage/FileStorage/XmlPlugin/Contract.cs (hunk 2 above)
Comment (Zhunusau) [reply lab-file-storage/FileStorage/XmlPlugin/Contract.cs]: renamed class to MyFileContract

File: lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs (hunk 5 above)
Comment (Zhunusau) [reply lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs]: rename method

File: lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs (hunk 6 above)
Comment (Zhunusau) [reply lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs]: made variable static

File: lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs (hunk 8 above)
Comment (Zhunusau) [reply lab-file-storage/FileStorage/FileStorage/BLL/Services/PluginManager.cs]: merge "if" in single one

File: lab-file-storage/FileStorage/FileStorage/BLL/Services/StorageService.cs (hunk 9 above)
Comment (Zhunusau) [reply lab-file-storage/FileStorage/FileStorage/BLL/Services/StorageService.cs]: made List<>

File: lab-file-storage/FileStorage/FileStorage/BLL/Models/User.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    public class User
+    {
+        public string Login { get; set; }
+        public string Password { get; set; }
+        public DateTime CreationDate { get; set; }
+        public long StorageUsed { get; set; }
+
+        public override string ToString()
+        {
+            return $" login: {Login}\n creation Date: {CreationDate:yyyy-MM-dd}\n storage used: {StorageUsed}";
```
Comment (a-yandulski) [inline lab-file-storage/FileStorage/FileStorage/BLL/Models/User.cs]: Add units for StorageUsed

File: lab-file-storage/FileStorage/FileStorage/BLL/Services/StorageService.cs [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+        {
+            var file = _files.FirstOrDefault(f => f.Key == fileName);
+            return file.Value?.ToString() ?? "File not found";
+        }
+
+        public string Upload(string pathFile)
+        {
+            var file = new FileInfo(pathFile);
+            if (file.Length > this.MaxFile)
+                return $"Maximum file size exceeded. Maximum file size {MaxFile/1000} kb";
```
[the remark this answers is not in the document]
Comment (Zhunusau) [reply lab-file-storage/FileStorage/FileStorage/BLL/Services/StorageService.cs]: fixed errors, added static method CalculateFileSize()

File: lab-file-storage/FileStorage/FileStorage/BLL/Services/StorageService.cs [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+            if (file.Length + this.UsedSize > this.MaxSize)
+                return $"Maximum storage size exceeded. Maximum storage size {MaxSize/1000000} Mb";
```
[the remark this answers is not in the document]
Comment (Zhunusau) [reply lab-file-storage/FileStorage/FileStorage/BLL/Services/StorageService.cs]: fixed

File: lab-file-storage/FileStorage/FileStorage/BLL/Models/MyFile.cs [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+        public string Name { get; set; }
+        public string Extension { get; set; }
+        public long Size { get; set; }
+        public string HashCode { get; set; }
+        public DateTime CreationDate { get; set; }
+        public int Downloads { get; set; }
+
+        public override string ToString()
+        {
+            return $" name: {Name}\n extension: {Extension}\n file size: {Size} b\n creation date: {CreationDate:yyyy-MM-dd}\n downloads: {Downloads}";
```
[the remark this answers is not in the document]
Comment (Zhunusau) [reply lab-file-storage/FileStorage/FileStorage/BLL/Models/MyFile.cs]: fixed

File: lab-file-storage/FileStorage/FileStorage/Commands/CommandManager.cs [hunk 16]
```diff
@@ -6,17 +6,23 @@
...
+        }
 
         public string FileDownload(DownloadOptions options)
         {
             if (!string.IsNullOrEmpty(options.FileName) && !string.IsNullOrEmpty(options.Path))
             {
                 return _storageService.Download(options.FileName, options.Path);
             }
-            return "File not found or directory does not exist";
+            return "Сommand entered incorrectly";
```
[the remark this answers is not in the document]
Comment (Zhunusau) [reply lab-file-storage/FileStorage/FileStorage/Commands/CommandManager.cs]: added const string Message

File: lab-file-storage/FileStorage/JsonPlugin/JsonSerializer.cs [hunk 17]
```diff
@@ -33,7 +33,7 @@ public void Export(Dictionary<string, MyFile> dictionary, string path)
             }
             catch (Exception e)
             {
-                throw new SerializationException("Erron in JsonPlagin " + e.Message);
+                throw new SerializationException("Erron in JsonPlagin. " + e.Message);
```
[the remark this answers is not in the document]
Comment (Zhunusau) [reply lab-file-storage/FileStorage/JsonPlugin/JsonSerializer.cs]: rename JsonPlugin

**Comment by a-yandulski** [comment]
Main functionality works as expected. However, the code still has lots of issues.
PR is approved.
Can be merged to master.

**Pull request merged**